onnxruntime/core/optimizer/computation_reduction.cc

```diff
@@ -0,0 +1,271 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/common/safeint.h"
+#include "core/graph/graph_utils.h"
+#include "core/optimizer/initializer.h"
+#include "core/optimizer/utils.h"
+#include "core/optimizer/computation_reduction.h"
+
+using namespace ONNX_NAMESPACE;
+using namespace ::onnxruntime::common;
+namespace onnxruntime {
+typedef std::function<Status(Graph&, Node&, Node&)> Handler;
+
+static const int GATHERND_BATCH_DIM = 1;
+
+static bool IsLeadingDimsEqual(const TensorShapeProto* input_shape, const TensorShapeProto* output_shape,
+                               int num_dim_to_check) {
+  ORT_ENFORCE(output_shape->dim_size() >= num_dim_to_check && input_shape->dim_size() >= num_dim_to_check);
+
+  for (int i = 0; i < num_dim_to_check; i++) {
+    auto& output_dim = output_shape->dim(i);
+    auto& input_dim = input_shape->dim(i);
+    if (output_dim.has_dim_value() && input_dim.has_dim_value()) {
+      if (output_dim.dim_value() != input_dim.dim_value()) {
+        return false;
+      }
+    } else if (output_dim.has_dim_param() && input_dim.has_dim_param()) {
+      if (output_dim.dim_param() != input_dim.dim_param()) {
+        return false;
+      }
+    } else {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+static int GetValidInputForGatherND(const Node& target_node) {
```

Comment:
>target_node [](start = 48, length = 11)

What is `target_node`? Consumer of `GatherND`?

Reply:
Yes you are right. I will add a comment for it.